File: 1007. Minimum Domino Rotations For Equal Row.go
// https://leetcode.com/problems/minimum-domino-rotations-for-equal-row/

package leetcode

import "math"

func minDominoRotations(A []int, B []int) int {
	if len(A) == 0 {
		return 0
	}

	a1, a2 := countRotations(A[0], A, B)
	b1, b2 := countRotations(B[0], A, B)
	result := min2(a1, a2, b1, b2)
	if result == math.MaxInt64 {
		return -1
	}
	return result
}

func countRotations(val int, A []int, B []int) (int, int) {
	aCount, bCount := 0, 0
	for i := range A {
		if A[i] != val {
			if B[i] == val {
				aCount++
			} else {
				aCount = math.MaxInt64
				break
			}
		}
	}

	for i := range B {
		if B[i] != val {
			if A[i] == val {
				bCount++
			} else {
				bCount = math.MaxInt64
				break
			}
		}
	}

	return aCount, bCount
}

func min2(nums ...int) int {
	result := math.MaxInt64
	for _, n := range nums {
		if n < result {
			result = n
		}
	}
	return result
}

File: 1007. Minimum Domino Rotations For Equal Row.js
// https://leetcode.com/problems/minimum-domino-rotations-for-equal-row/

/**
 * @param {number[]} A
 * @param {number[]} B
 * @return {number}
 */
var minDominoRotations = function(A, B) {
    if(!A.length) return 0;

    const count = (v) => {
        let [aCount, bCount] = [0, 0];
        // convert A[x] to v
        for(let i=0; i<A.length; i++){
            if(A[i] !== v){
                if(B[i] === v) aCount++;
                else aCount = Infinity;
            }
        }

        // convert B[x] to v
        for(let i=0; i<B.length; i++){
            if(B[i] !== v){
                if(A[i] === v) bCount++;
                else bCount = Infinity;
            }
        }

        return [aCount, bCount];
    };

    const result = Math.min(...count(A[0]), ...count(B[0]));

    return result === Infinity ? -1 : result;
}